plugins/org.eclipse.glsp.server/src/org/eclipse/glsp/server/features/core/model/ModelSubmissionHandler.java

```diff
@@ -55,28 +57,28 @@
     * </p>
     *
     * @param modelState The model state to submit.
+    * @param causedBy   The action/operation that caused the model update
     * @return A list of actions to be processed in order to submit the model.
     */
-   public List<Action> submitModel(final GModelState modelState) {
+   public List<Action> submitModel(final GModelState modelState, final Action causedBy) {
       modelFactory.createGModel(modelState);
       modelState.getRoot().setRevision(modelState.getRoot().getRevision() + 1);
       DiagramConfiguration diagramConfiguration = diagramConfigurationRegistry.get(modelState);
-      if (diagramConfiguration.getLayoutKind() == ServerLayoutKind.AUTOMATIC) {
-         layoutEngine.layout(modelState);
-      }
-      if (diagramConfiguration.needsClientLayout()) {
+      boolean needsClientLayout = diagramConfiguration.needsClientLayout();
+      if (needsClientLayout) {
          synchronized (modelLock) {
             return Arrays.asList(new RequestBoundsAction(modelState.getRoot()),
-               new SetDirtyStateAction(modelState.isDirty()));
+               new SetDirtyStateAction(modelState.isDirty(), causedBy));
          }
       }
-      return submitModelDirectly(modelState);
+      return submitModelDirectly(modelState, needsClientLayout ? null : causedBy);
```

Comment:
It seems like it should not be possible for `needsClientLayout` to be true here. If it were, we would already have returned.